presto-main/src/main/java/io/prestosql/sql/planner/ScopeAware.java

```diff
@@ -0,0 +1,182 @@
+ *
+ * For example, given
+ *
+ * <pre>SELECT t.a, a FROM (VALUES 1) t(a)</pre>
+ *
+ * "t.a" and "a" are considered equal because they reference the same field of "t"
+ *
+ * Limitation: the expressions in the following query are currently not considered equal to each other, even though they refer to the same field from the same table or
+ * named query "t". This is because we currently don't assign identity to table references. (TODO: implement this)
+ *
+ * <pre>SELECT (SELECT t.a FROM t), (SELECT a FROM t)</pre>
+ */
+public class ScopeAware<T extends Node>
+{
+    private final Analysis analysis;
+    private final Scope scope;
+    private final T node;
+    private final int hash;
+
+    private ScopeAware(Analysis analysis, Scope scope, T node)
+    {
+        requireNonNull(scope, "scope is null");
+
+        this.scope = findQueryBoundaryScope(scope);
+        this.analysis = requireNonNull(analysis, "analysis is null");
+        this.node = requireNonNull(node, "node is null");
+        this.hash = hash(node);
+    }
+
+    private static Scope findQueryBoundaryScope(Scope scope)
+    {
+        Optional<Scope> parent = scope.getLocalParent();
+        while (parent.isPresent()) {
+            scope = parent.get();
+            parent = scope.getLocalParent();
+        }
+        return scope;
+    }
+
+    public static <T extends Node> ScopeAware<T> scopeAwareKey(T node, Analysis analysis, Scope scope)
+    {
+        return new ScopeAware<>(analysis, scope, node);
+    }
+
+    public T getNode()
+    {
+        return node;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return hash;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        ScopeAware<T> other = (ScopeAware<T>) o;
+        checkArgument(this.scope == other.scope, "Expressions must be in the same local scope");
+
+        return equal(node, other.node);
+    }
+
+    @Override
+    public String toString()
+    {
+        return "ScopeAware(" + node + ")";
+    }
+
+    private boolean equal(Node left, Node right)
+    {
+        if (left instanceof Expression && right instanceof Expression) {
+            Expression leftExpression = (Expression) left;
+            Expression rightExpression = (Expression) right;
+            if (analysis.isColumnReference(leftExpression) && analysis.isColumnReference(rightExpression)) {
+                ResolvedField leftField = analysis.getResolvedField(leftExpression);
+                ResolvedField rightField = analysis.getResolvedField(rightExpression);
+
+                Scope leftScope = leftField.getScope();
+                Scope rightScope = rightField.getScope();
+
+                // For subqueries of the query associated with the current expression, compare by syntax
```

Comment:
It would be great to have some test coverage for these edge cases. I'm not sure how feasible writing tests for `ScopeAware` would be. However, tests would show practical examples of same expressions